sound/soc/sof/trace.c

```diff
@@ -13,6 +13,216 @@
 #include "sof-priv.h"
 #include "ops.h"
 
+#define TRACE_FILTER_MAX_IPC_ELEMENTS_PER_ENTRY 4
+
+static int trace_filter_append_elem(int32_t key, int32_t value,
+				    struct sof_ipc_trace_filter_elem *elem_list,
+				    size_t capacity, size_t *counter)
+{
+	if (*counter >= capacity)
+		return -ENOMEM;
+
+	elem_list[*counter].key = key;
+	elem_list[*counter].value = value;
+	++*counter;
+
+	return 0;
+}
+
+static int trace_filter_parse_entry(struct snd_sof_dev *sdev, const char *line,
+				    struct sof_ipc_trace_filter_elem *elem,
+				    size_t capacity, size_t *counter)
+{
+	int len = strlen(line);
+	size_t cnt = *counter;
+	int log_level;
+	int uuid_id;
+	int pipe_id;
+	int comp_id;
+	int read;
+	int ret;
+
+	ret = sscanf(line, " %d %x %d %d %n", &log_level, &uuid_id, &pipe_id, &comp_id, &read);
+	if (ret != 4 || read != len) {
+		dev_err(sdev->dev, "error: invalid trace filter entry '%s'\n",
+			line);
+		return -EINVAL;
+	}
+
+	if (uuid_id > 0) {
+		ret = trace_filter_append_elem(SOF_IPC_TRACE_FILTER_ELEM_UUID,
+					       uuid_id, elem, capacity, &cnt);
+		if (ret)
+			return ret;
+	}
+	if (pipe_id >= 0) {
+		ret = trace_filter_append_elem(SOF_IPC_TRACE_FILTER_ELEM_PIPE,
+					       pipe_id, elem, capacity, &cnt);
+		if (ret)
+			return ret;
+	}
+	if (comp_id >= 0) {
+		ret = trace_filter_append_elem(SOF_IPC_TRACE_FILTER_ELEM_COMP,
+					       comp_id, elem, capacity, &cnt);
+		if (ret)
+			return ret;
+	}
+
+	ret = trace_filter_append_elem(SOF_IPC_TRACE_FILTER_ELEM_LEVEL |
+				       SOF_IPC_TRACE_FILTER_ELEM_FIN,
+				       log_level, elem, capacity, &cnt);
+	if (ret)
+		return ret;
+
+	/* update counter only when parsing whole entry passed */
+	*counter = cnt;
+
+	return len;
+}
+
+static int trace_filter_parse(struct snd_sof_dev *sdev, char* string,
+			      size_t len, size_t *out_elem_cnt,
+			      struct sof_ipc_trace_filter_elem **out)
+{
+	const char entry_delimiter = ';';
+	char *entry = string;
+	size_t capacity = 0;
+	char *entry_end;
+	size_t cnt = 0;
+	int entry_len;
+
+	/* calculate capacity for the worst case scenario */
```

Comment:
saying worst case scenario would raise eyebrows. What we're doing here is calculating the required capacity for the actual number fully formatted entry strings isnt it?

Reply:
rephrased